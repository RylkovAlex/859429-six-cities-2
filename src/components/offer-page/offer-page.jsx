import React, {useEffect} from "react";
import PropTypes from "prop-types";
import {
  reviewPropTypes,
  offerCardPropTypes
} from "../../prop-types/prop-types";
import {connect} from "react-redux";
import OffersList, {ListType} from "../offers-list/offers-list.jsx";
import ReviewsList from "../reviews-list/reviews-list.jsx";
import Header from "../header/header.jsx";
import ActionCreator, {
  Operation
} from "../../redux/actions/action-creator/action-creator";
import ReviewForm from "../review-form/review-form.jsx";
import {getNearbyCards, getMapConfig} from "../../redux/selectors/selectors";
import MapComponent from "../map-component/map-component.jsx";

const OfferPage = (props) => {
  const {
    reviews,
    offers,
    isAuthorized,
    changeActiveCard,
    postFavorite,
    isFetching,
    nearbyCards,
    loadReviews,
    isReviewsLoading,
    isReviewsLoadingError,
    activeCard
  } = props;
  const hotelId = +props.match.params.id;
  const card = offers.find((offer) => +offer.id === hotelId);
  const {
    id,
    images,
    title,
    isPremium,
    isFavorite,
    rating,
    type,
    bedrooms,
    maxAdults,
    price,
    goods,
    host,
    description
  } = card;
  const ratingStyle = {
    width: `${rating * 20}%`
  };

  useEffect(() => {
    loadReviews(id);
  }, []);

  return (
    <div className="page">
      <Header />
      <main className="page__main page__main--property">
        <section className="property">
          <div className="property__gallery-container container">
            <div className="property__gallery">
              {images.map((src, i) => {
                return (
                  <div
                    className="property__image-wrapper"
                    key={`property__image - ${i}`}
                  >
                    <img
                      className="property__image"
                      src={src}
                      alt="Photo studio"
                    />
                  </div>
                );
              })}
            </div>
          </div>
          <div className="property__container container">
            <div className="property__wrapper">
              <div className="property__mark">
                {isPremium ? <span>Premium</span> : ``}
              </div>
              <div className="property__name-wrapper">
                <h1 className="property__name">{title}</h1>
                <button
                  className={`property__bookmark-button button ${
                    isFavorite ? `place-card__bookmark-button--active` : ``
                  }`}
                  type="button"
                  disabled={isFetching}
                  onClick={() => postFavorite(id, !isFavorite)}
                >
                  <svg
                    className="property__bookmark-icon place-card__bookmark-icon"
                    width="31"
                    height="33"
                  >
                    <use xlinkHref="#icon-bookmark"></use>
                  </svg>
                  <span className="visually-hidden">To bookmarks</span>
                </button>
              </div>
              <div className="property__rating rating">
                <div className="property__stars rating__stars">
                  <span style={ratingStyle}></span>
                  <span className="visually-hidden">Rating</span>
                </div>
                <span className="property__rating-value rating__value">
                  {rating}
                </span>
              </div>
              <ul className="property__features">
                <li className="property__feature property__feature--entire">
                  {type}
                </li>
                <li className="property__feature property__feature--bedrooms">
                  {bedrooms} Bedrooms
                </li>
                <li className="property__feature property__feature--adults">
                  Max {maxAdults} adults
                </li>
              </ul>
              <div className="property__price">
                <b className="property__price-value">&euro;{price}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>
              <div className="property__inside">
                <h2 className="property__inside-title">What&apos;s inside</h2>
                <ul className="property__inside-list">
                  {goods.map((good, i) => (
                    <li
                      className="property__inside-item"
                      key={`inside-item-${i}`}
                    >
                      {good}
                    </li>
                  ))}
                </ul>
              </div>
              <div className="property__host">
                <h2 className="property__host-title">Meet the host</h2>
                <div className="property__host-user user">
                  <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                    <img
                      className="property__avatar user__avatar"
                      src={`/${host.avatarUrl}`}
                      width="74"
                      height="74"
                      alt="Host avatar"
                    />
                  </div>
                  <span className="property__user-name">{host.name}</span>
                  <span className="property__user-status">
                    {host.isPro ? `Pro` : ``}
                  </span>
                </div>
                <div className="property__description">
                  <p className="property__text">{description}</p>
                </div>
              </div>
              {/* ОТЗЫВЫ */}
              <section className="property__reviews reviews">
                <h2 className="reviews__title">
                  Reviews &middot; <span className="reviews__amount">1</span>
                </h2>
                {/* Список отзывов */}
                {isReviewsLoadingError && (
                  <p>
                    Connection problem, can`t load reviews! To try again refresh
                    page!
                  </p>
                )}
                {isReviewsLoading ? (
                  <p>Loading...</p>
                ) : (
                  <ReviewsList reviews={reviews} />
                )}
                {/* Форма для отправки нового отзыва */}
                {isAuthorized && <ReviewForm hotelId={hotelId} />}
              </section>
            </div>
          </div>
          {/* КАРТА */}
          <section className="property__map map">
            <MapComponent
              config={getMapConfig(nearbyCards)}
              activePointId={activeCard}
            />
          </section>
        </section>

        {/* Похожие */}
        <div className="container">
          <section className="near-places places">
            <h2 className="near-places__title">
              Other places in the neighbourhood
            </h2>
            <OffersList
              offerCards={nearbyCards}
              listType={ListType.NearbyList}
              handleCardHover={changeActiveCard}
              handleBookmarkClick={postFavorite}
            ></OffersList>
          </section>
        </div>
      </main>
    </div>
  );
};

OfferPage.propTypes = {
  activeCard: PropTypes.number.isRequired,
  offers: PropTypes.arrayOf(PropTypes.shape(offerCardPropTypes)).isRequired,
  nearbyCards: PropTypes.arrayOf(PropTypes.shape(offerCardPropTypes)),
  reviews: PropTypes.arrayOf(PropTypes.shape(reviewPropTypes)),
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string
    })
  }).isRequired,
  changeActiveCard: PropTypes.func.isRequired,
  postFavorite: PropTypes.func.isRequired,
  loadReviews: PropTypes.func.isRequired,
  isAuthorized: PropTypes.bool,
  isReviewsLoading: PropTypes.bool.isRequired,
  isReviewsLoadingError: PropTypes.bool.isRequired,
  isFetching: PropTypes.bool
};

export {OfferPage};

const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    activeCard: state.activeCard,
    nearbyCards: getNearbyCards(ownProps.match.params.id, state),
    offers: state.allOffers,
    reviews: state.reviews,
    isAuthorized: !!state.user,
    isReviewsLoading: state.isReviewsLoading,
    isReviewsLoadingError: state.isReviewsLoadingError,
    isFetching: state.isFetching,
    isFetchError: state.isFetchError
  });

const mapDispatchToProps = (dispatch) => ({
  changeActiveCard: (cardId) => dispatch(ActionCreator.changeActiveCard(cardId)),
  postFavorite: (cardId, status) =>
    dispatch(Operation.postFavorite(cardId, status)),
  loadReviews: (cardId) => dispatch(Operation.loadReviews(cardId))
});

export default connect(mapStateToProps, mapDispatchToProps)(OfferPage);
