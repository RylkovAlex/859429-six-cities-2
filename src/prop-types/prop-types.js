import PropTypes from 'prop-types';

export const offerCardPropTypes = {
  city: PropTypes.shape({
    name: PropTypes.string.isRequired,
    location: PropTypes.shape({
      latitide: PropTypes.number,
      longitude: PropTypes.number,
      zoom: PropTypes.number,
    }),
  }),
  id: PropTypes.number.isRequired,
  previewImage: PropTypes.string.isRequired,
  images: PropTypes.arrayOf(PropTypes.string),
  title: PropTypes.string.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  isPremium: PropTypes.bool.isRequired,
  rating: PropTypes.number.isRequired,
  type: PropTypes.string.isRequired,
  bedrooms: PropTypes.number.isRequired,
  maxAdults: PropTypes.number.isRequired,
  price: PropTypes.number.isRequired,
  goods: PropTypes.arrayOf(PropTypes.string),
  host: PropTypes.shape({
    id: PropTypes.number.isRequired,
    isPro: PropTypes.bool,
    name: PropTypes.stringisRequired,
    avatarUrl: PropTypes.string,
  }),
  description: PropTypes.string,
  location: PropTypes.shape({
    latitide: PropTypes.number,
    longitude: PropTypes.number,
    zoom: PropTypes.number,
  }),
};

export const reviewPropTypes = {
  id: PropTypes.number.isRequired,
  user: PropTypes.shape({
    id: PropTypes.number.isRequired,
    isPro: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
    avatarUrl: PropTypes.string.isRequired,
  }),
  rating: PropTypes.number.isRequired,
  comment: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
};

export const cityPropTypes = {
  name: PropTypes.string.isRequired,
  location: PropTypes.shape({
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
    zoom: PropTypes.number.isRequired,
  }),
};

export const offerCardForTests = {
  city: {
    name: ``,
    location: {
      latitude: 1,
      longitude: 1,
      zoom: 1,
    },
  },
  id: 0,
  previewImage: ``,
  images: [``],
  title: ``,
  isFavorite: false,
  isPremium: false,
  rating: 0,
  type: ``,
  bedrooms: 0,
  maxAdults: 0,
  price: 0,
  goods: [``],
  host: {
    id: 0,
    isPro: false,
    name: ``,
    avatarUrl: ``,
  },
  description: ``,
  location: {
    latitude: 1,
    longitude: 1,
    zoom: 1,
  },
};

export const reviewForTests = {
  id: 0,
  user: {
    id: 0,
    isPro: false,
    name: ``,
    avatarUrl: ``,
  },
  rating: 0,
  comment: ``,
  date: `Thu Nov 07 2019 13:56:54 GMT+0300 (Москва, стандартное время)`,
};
