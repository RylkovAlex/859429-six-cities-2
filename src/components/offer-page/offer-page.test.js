import React from 'react';
import renderer from 'react-test-renderer';
import {OfferPage} from './offer-page';
import {offerCardForTests} from '../../prop-types/prop-types';

/* const copmponentsToMock = {
  OffersList: `../offers-list/offers-list.jsx`,
  ReviewsList: `../reviews-list/reviews-list.jsx`,
  Header: `../header/header.jsx`,
  OfferCard: `../offer-card/offer-card.jsx`,
  ReviewForm: `../review-form/review-form.jsx`,
}; */

// TODO: не работает..
/* const modules = Object.values(copmponentsToMock);
for (let i = 0; i < modules.length; i++) {
  jest.mock(modules[i], () => jest.fn().mockReturnValue(null));
} */

jest.mock(`../map-component/map-component.jsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`../../redux/selectors/selectors`, () => ({
  __esModule: true,
  getMapConfig: jest.fn().mockReturnValue(null),
}));
jest.mock(`../header/header.jsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`../review-form/review-form.jsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`../offers-list/offers-list.jsx`, () => ({
  __esModule: true,
  ListType: {
    NearbyList: ``
  },
  OfferList: () => null,
  default: () => null,
}));

it(`OfferPage correctly renders after relaunch`, () => {
  const props = {
    activeCard: 0,
    loadReviews: jest.fn(),
    isReviewsLoading: false,
    isReviewsLoadingError: false,
    offers: [offerCardForTests],
    reviews: [],
    match: {
      params: {
        id: `0`,
      }
    },
    changeActiveCard: jest.fn(),
    postFavorite: jest.fn(),
    isAuthorized: true,
  };

  const tree = renderer
    .create(<OfferPage {...props}/>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
