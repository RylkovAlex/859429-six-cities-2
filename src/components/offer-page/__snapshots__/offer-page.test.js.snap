// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferPage correctly renders after relaunch 1`] = `
<div
  className="page"
>
  <div />
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/room.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/apartment-01.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/apartment-02.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/apartment-03.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/studio-01.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="/img/apartment-01.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__mark"
          >
            
          </div>
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Beautiful & luxurious studio at great location
            </h1>
            <button
              className="property__bookmark-button button "
              onClick={[Function]}
              type="button"
            >
              <svg
                className="property__bookmark-icon place-card__bookmark-icon"
                height="33"
                width="31"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "56%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              2.8
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              apartment
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              3
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              4
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              20
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Heating
              </li>
              <li
                className="property__inside-item"
              >
                Kitchen
              </li>
              <li
                className="property__inside-item"
              >
                Cable TV
              </li>
              <li
                className="property__inside-item"
              >
                Washing machine
              </li>
              <li
                className="property__inside-item"
              >
                Coffee machine
              </li>
              <li
                className="property__inside-item"
              >
                Dishwasher
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="//img/avatar-angelina.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews · 
              <span
                className="reviews__amount"
              >
                1
              </span>
            </h2>
            <div />
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      >
        <div />
      </section>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
      </section>
    </div>
  </main>
</div>
`;
